clients/src/main/java/org/apache/kafka/common/utils/ByteBufferOutputStream.java

```diff
@@ -40,18 +42,49 @@ public void write(int b) {
 
     public void write(byte[] bytes, int off, int len) {
         if (buffer.remaining() < len)
-            expandBuffer(buffer.capacity() + len);
+            expandBuffer(buffer.position() + len);
         buffer.put(bytes, off, len);
     }
 
+    public void write(ByteBuffer buffer) {
+        if (buffer.hasArray())
+            write(buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.remaining());
+        else {
+            int pos = buffer.position();
+            for (int i = pos, limit = buffer.remaining() + pos; i < limit; i++)
+                write(buffer.get(i));
+        }
+    }
+
     public ByteBuffer buffer() {
         return buffer;
     }
 
+    public int position() {
+        return buffer.position();
+    }
+
+    public int capacity() {
+        return buffer.capacity();
+    }
+
+    public int limit() {
+        return buffer.limit();
+    }
+
+    public void position(int position) {
+        if (position > buffer.limit())
+            expandBuffer(position);
+        buffer.position(position);
+    }
+
     private void expandBuffer(int size) {
         int expandSize = Math.max((int) (buffer.capacity() * REALLOCATION_FACTOR), size);
         ByteBuffer temp = ByteBuffer.allocate(expandSize);
         temp.put(buffer.array(), buffer.arrayOffset(), buffer.position());
+
+        // reset the old buffer's position so that
+        buffer.position(initialPosition);
```

Comment:
Embarrassed I didn't catch that since I thought about it when implementing `write(ByteBuffer)`.